.github/workflows/ci.yml

```diff
@@ -6,182 +6,11 @@ on:
-        if: runner.os == 'Windows'
-        id: windows-dlls-cache
-        uses: actions/cache@v4
-        with:
-          path: external/dlls-${{ matrix.target.cpu }}
-          key: 'dlls-${{ matrix.target.cpu }}'
-
-      - name: Install DLLs dependencies (Windows)
-        if: >
-          steps.windows-dlls-cache.outputs.cache-hit != 'true' &&
-          runner.os == 'Windows'
-        run: |
-          mkdir -p external
-          curl -L "https://nim-lang.org/download/windeps.zip" -o external/windeps.zip
-          7z x -y external/windeps.zip -oexternal/dlls-${{ matrix.target.cpu }}
-
-      - name: Path to cached dependencies (Windows)
-        if: >
-          runner.os == 'Windows'
-        run: |
-          echo "${{ github.workspace }}/external/dlls-${{ matrix.target.cpu }}" >> $GITHUB_PATH
-
-      - name: Derive environment variables
-        run: |
-          if [[ '${{ matrix.target.cpu }}' == 'amd64' ]]; then
-            PLATFORM=x64
-          elif [[ '${{ matrix.target.cpu }}' == 'arm64' ]]; then
-            PLATFORM=arm64
-          else
-            PLATFORM=x86
-          fi
-          echo "PLATFORM=$PLATFORM" >> $GITHUB_ENV
-
-          ncpu=
-          MAKE_CMD="make"
-          case '${{ runner.os }}' in
-          'Linux')
-            ncpu=$(nproc)
-            ;;
-          'macOS')
-            ncpu=$(sysctl -n hw.ncpu)
-            ;;
-          'Windows')
-            ncpu=$NUMBER_OF_PROCESSORS
-            MAKE_CMD="mingw32-make"
-            ;;
-          esac
-          [[ -z "$ncpu" || $ncpu -le 0 ]] && ncpu=1
-          echo "ncpu=$ncpu" >> $GITHUB_ENV
-          echo "MAKE_CMD=${MAKE_CMD}" >> $GITHUB_ENV
-
-      - name: Build Nim and Nimble
-        run: |
-          curl -O -L -s -S https://raw.githubusercontent.com/status-im/nimbus-build-system/master/scripts/build_nim.sh
-          env MAKE="${MAKE_CMD} -j${ncpu}" ARCH_OVERRIDE=${PLATFORM} NIM_COMMIT=${{ matrix.branch }} \
-            QUICK_AND_DIRTY_COMPILER=1 QUICK_AND_DIRTY_NIMBLE=1 CC=gcc \
-            bash build_nim.sh nim csources dist/nimble NimBinaries
-          echo '${{ github.workspace }}/nim/bin' >> $GITHUB_PATH
-
-      - name: Use gcc 14
-        # Should be removed when ubuntu-latest is 26.04
-        if : ${{ matrix.target.os == 'linux-gcc-14' }}
-        run: |
-          # Add GCC-14 to alternatives
-          sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-14 14
-          # Set GCC-14 as the default
-          sudo update-alternatives --set gcc /usr/bin/gcc-14
-
-      - name: Run tests
-        run: |
-          nim --version
-          nimble --version
-          gcc --version
-          nimble install -y --depsOnly
-
-          env NIMLANG=c nimble test
-
-          nimble install chronos
-          env NIMLANG=c nimble testChronos || true
+    uses: status-im/nimbus-common-workflow/.github/workflows/common.yml@main
+    with:
+      test-command: |
```

Comment:
can't this be a normal step? ie why does it have to be inside the action?

Reply:
the normal way of developing actions is that "setup" steps like the common workflow "prepare" the environment but don't take parameters to actually run user tasks - those should live in separate steps - see the example here: https://github.com/jiro4989/setup-nim-action?tab=readme-ov-file#basic-usage

also, not all project will have the exact same steps for their "run tests" section".. in particular, `nimble install -y --depsOnly` should probably not be part of the common workflow since it's on its way out for many projects and already performs environment-altering actions from which it's hard to recover to a "pristine" testing environment

Reply:
What you've linked is a Github *Action*. The stuff implemented here is a Github *Workflow*.
An action is a step in a larger workflow.

What this workflow offers and what an action doesn't (and shouldn't) is a unified control over Nim branches and OSes which are tested.
It's not just "bleh, I need to make 40+ PRs" when something changes. Currently we have "mostly the same, but slightly different (not on purpose!)" workflows in each project. If you consider those projects as a part of a bigger picture, until this change we couldn't claim that everything is tested in the same environment(s).


The way the common workflow is implemented _vastly_ improves the situation we had until now.
And it keeps things easily maintainable (one easily understandable yaml file), not just for me, but for anyone who wants to make a change.
I don't want to write nor maintain an equivalent of [this](https://github.com/jiro4989/setup-nim-action/blob/master/install_nim.sh) from your example.



> in particular, `nimble install -y --depsOnly` should probably not be part of the common workflow since it's on its way out for many projects

This line is currently present in ALL the packages that are receiving this update.
Maybe it is on its way out, but nobody got the memo. 
(This is just another example (among many) why the current situation is bad)

If/when it will be "really out", it will be just one PR away from removing it everywhere. (Again, if some package really needs it, it can add it back via custom `test-command`)
If you want, I can remove it already and see if anything breaks.

Reply:
> The stuff implemented here is a Github Workflow.

Is it possible then to add additional steps, when those steps use other github actions (like [upload](https://github.com/actions/upload-artifact?tab=readme-ov-file#breaking-changes))? ie how would we publish benchmark results from something like bncurve/stint etc?